crates/nu-command/src/network/port.rs

```diff
@@ -121,25 +120,200 @@ fn get_free_port(
+            // - Casts between slice DSTs preserve the length metadata:
+            //     https://github.com/rust-lang/unsafe-code-guidelines/issues/288
+            //     https://github.com/rust-lang/reference/pull/1417
+            // - Casting to `*mut TcpTable` preserves that metadata for our DST.
+            // - `Box::from_raw` takes ownership and will free via the same allocator.
+            let table = unsafe {
+                let ptr = ptr::slice_from_raw_parts_mut(ptr, header.dwNumEntries as usize);
+                Box::from_raw(ptr as *mut TcpTable)
+            };
+
+            Ok(table)
+        }
+    }
+
+    /// Find an open port by checking the TCP table.
+    ///
+    /// On Windows, it is possible to bind to the same port multiple times if it was not
+    /// originally bound as an exclusive port[^so].
+    /// The Rust implementation of [`TcpListener::bind`] currently does not enforce exclusive
+    /// binding, which means the same port can be bound more than once.  
+    /// Because of this, we cannot simply try binding to a port to check if it is free.  
+    /// Instead, we query the [TCP table](https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-gettcptable2)
+    /// to see which ports are already in use and then pick one that is not listed.
+    ///
+    /// [^so]: <https://docs.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse>
+    #[cfg(windows)]
+    pub fn search_port_in_range(
+        range: Spanned<RangeInclusive<u16>>,
+        call_span: Span,
+    ) -> Result<u16, ShellError> {
+        use std::collections::HashSet;
+
+        let table = TcpTable::new()
+            .map_err(|err| {
+                (
+                    err,
+                    match err {
+                        NO_ERROR => unreachable!("handled as Ok variant"),
+                        ERROR_INSUFFICIENT_BUFFER => "The buffer for TcpTable is not large enough",
+                        ERROR_INVALID_PARAMETER => "SizePointer was null or not writable",
+                        ERROR_NOT_SUPPORTED => "GetTcpTable2 is not supported on this OS",
+                        _ => "Unexpected error code from GetTcpTable2",
+                    },
+                )
+            })
+            .map_err(|(err, msg)| {
+                ShellError::Io(IoError::new_with_additional_context(
+                    std::io::Error::from_raw_os_error(err.0 as i32),
+                    call_span,
+                    None,
+                    msg,
+                ))
+            })?;
+
+        let used_ports: HashSet<u16> = table
+            .table
+            .iter()
+            .map(|row| row.dwLocalPort as u16)
+            .map(|raw| {
+                // Convert from network byte order to host byte order.
+                // SAFETY: `raw` is the exact value returned by the API for a port.
+                unsafe { ntohs(raw) }
+            })
```

Comment:
Those two `map`s can be collapsed into one.

`ntohs` being unsafe is overly cautious as an FFI call. It being effectively `u16::from_be` and an operation purely on pass-by-value.